cupyx/scipy/sparse/csr.py

```diff
@@ -357,3 +372,223 @@ def isspmatrix_csr(x):
+        else:
+            indptr = cupy.arange(0, nnz+1, sp.nnz, dtype=sp.indptr.dtype)
+    else:
+        indptr = sp.indptr.copy()
+        if n > sp_n:
+            indptr *= n
+
+    # out = sp * dn
+    cupy_multiply_by_dense()(sp.data, sp.indptr, sp.indices, sp_m, sp_n,
+                             dn, dn_m, dn_n, indptr, m, n, data, indices)
+
+    return csr_matrix((data, indices, indptr), shape=(m, n))
+
+
+def cupy_multiply_by_dense():
+    return cupy.ElementwiseKernel(
+        '''
+        raw S SP_DATA, raw I SP_INDPTR, raw I SP_INDICES,
+        int32 SP_M, int32 SP_N,
+        raw D DN_DATA, int32 DN_M, int32 DN_N,
+        raw I OUT_INDPTR, int32 OUT_M, int32 OUT_N
+        ''',
+        'O OUT_DATA, I OUT_INDICES',
+        '''
+        int i_out = i;
+        int _min = 0;
+        int _max = OUT_M - 1;
+        int m_out = (_min + _max) / 2;
+        while (_min < _max) {
+            if (i_out < OUT_INDPTR[m_out]) {
+                _max = m_out - 1;
+            }
+            else if (i_out >= OUT_INDPTR[m_out+1]) {
+                _min = m_out + 1;
+            }
+            else {
+                break;
+            }
+            m_out = (_min + _max) / 2;
+        }
+        int i_sp = i_out;
+        if (OUT_M > SP_M && SP_M == 1) {
+            i_sp -= OUT_INDPTR[m_out];
+        }
+        if (OUT_N > SP_N && SP_N == 1) {
+            i_sp /= OUT_N;
+        }
+        int n_out = SP_INDICES[i_sp];
+        if (OUT_N > SP_N && SP_N == 1) {
+            n_out = i_out - OUT_INDPTR[m_out];
+        }
+        int m_dn = m_out;
+        if (OUT_M > DN_M && DN_M == 1) {
+            m_dn = 0;
+        }
+        int n_dn = n_out;
+        if (OUT_N > DN_N && DN_N == 1) {
+            n_dn = 0;
+        }
+        OUT_DATA = (O)(SP_DATA[i_sp] * DN_DATA[n_dn + (DN_N * m_dn)]);
+        OUT_INDICES = n_out;
+        ''',
+        'cupy_multiply_by_dense'
+    )
+
+
+def multiply_by_csr(a, b):
+    a_m, a_n = a.shape
+    b_m, b_n = b.shape
+    if not (a_m == b_m or a_m == 1 or b_m == 1):
+        raise ValueError('inconsistent shape')
+    if not (a_n == b_n or a_n == 1 or b_n == 1):
+        raise ValueError('inconsistent shape')
+    m, n = max(a_m, b_m), max(a_n, b_n)
+    a_nnz = a.nnz * (m // a_m) * (n // a_n)
+    b_nnz = b.nnz * (m // b_m) * (n // b_n)
+    if a_nnz > b_nnz:
+        return multiply_by_csr(b, a)
+    c_nnz = a_nnz
+    dtype = functools.reduce(numpy.promote_types, (a.dtype, b.dtype))
+    c_data = cupy.empty(c_nnz, dtype=dtype)
+    c_indices = cupy.empty(c_nnz, dtype=a.indices.dtype)
+    if m > a_m:
+        if n > a_n:
+            c_indptr = cupy.arange(0, c_nnz+1, n, dtype=a.indptr.dtype)
+        else:
+            c_indptr = cupy.arange(0, c_nnz+1, a.nnz, dtype=a.indptr.dtype)
+    else:
+        c_indptr = a.indptr.copy()
+        if n > a_n:
+            c_indptr *= n
+    flags = cupy.zeros(c_nnz+1, dtype=a.indices.dtype)
```

Comment:
Up here the code is 99% same as multiply by dense, I think some of these checks could be done in the regular `multiply` function to avoid code repetition.

Reply:
Yes, they are almost identical, but there are a few differences, so I'd like to keep them separate for now.